src/libraries/System.Collections/src/System/Collections/Generic/ReadOnlySet.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Collections.ObjectModel
+{
+    /// <summary>Represents a read-only, generic set of values.</summary>
+    /// <typeparam name="T">The type of values in the set.</typeparam>
+    [DebuggerDisplay("Count = {Count}")]
+    public class ReadOnlySet<T> : IReadOnlySet<T>, ISet<T>, ICollection
+    {
+        /// <summary>The wrapped set.</summary>
+        private readonly ISet<T> _set;
+
+        /// <summary>Initializes a new instance of the <see cref="ReadOnlySet{T}"/> class that is a wrapper around the specified set.</summary>
+        /// <param name="set">The set to wrap.</param>
+        public ReadOnlySet(ISet<T> set)
+        {
+            ArgumentNullException.ThrowIfNull(set);
+            _set = set;
+        }
+
+        /// <summary>Gets an empty <see cref="ReadOnlySet{T}"/>.</summary>
+        public static ReadOnlySet<T> Empty { get; } = new ReadOnlySet<T>(new HashSet<T>());
+
+        /// <summary>Gets the set that is wrapped by this <see cref="ReadOnlySet{T}"/> object.</summary>
+        protected ISet<T> Set => _set;
+
+        /// <inheritdoc/>
+        public int Count => _set.Count;
+
+        /// <inheritdoc/>
+        public IEnumerator<T> GetEnumerator() =>
+            _set.Count == 0 ? ((IEnumerable<T>)Array.Empty<T>()).GetEnumerator() :
+            _set.GetEnumerator();
+
+        /// <inheritdoc/>
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <inheritdoc/>
+        public bool Contains(T item) => _set.Contains(item);
+
+        /// <inheritdoc/>
+        public bool IsProperSubsetOf(IEnumerable<T> other) => _set.IsProperSubsetOf(other);
+
+        /// <inheritdoc/>
+        public bool IsProperSupersetOf(IEnumerable<T> other) => _set.IsProperSupersetOf(other);
+
+        /// <inheritdoc/>
+        public bool IsSubsetOf(IEnumerable<T> other) => _set.IsSubsetOf(other);
+
+        /// <inheritdoc/>
+        public bool IsSupersetOf(IEnumerable<T> other) => _set.IsSupersetOf(other);
+
+        /// <inheritdoc/>
+        public bool Overlaps(IEnumerable<T> other) => _set.Overlaps(other);
+
+        /// <inheritdoc/>
+        public bool SetEquals(IEnumerable<T> other) => _set.SetEquals(other);
+
+        /// <inheritdoc/>
+        void ICollection<T>.CopyTo(T[] array, int arrayIndex) => _set.CopyTo(array, arrayIndex);
+
+        /// <inheritdoc/>
+        void ICollection.CopyTo(Array array, int index) => CollectionHelpers.CopyTo(_set, array, index);
+
+        /// <inheritdoc/>
+        bool ICollection<T>.IsReadOnly => true;
+
+        /// <inheritdoc/>
+        bool ICollection.IsSynchronized => false;
```

Comment:
Similar to `SyncRoot`, shouldn't this forward to the `ICollection` implementation if available?

Reply:
This is following what both ReadOnlyCollection and ReadOnlyDictionary do here.

Reply:
Do you know why? It seems arguably wrong in those cases too.

Of course, hopefully nobody's writing new code that uses this property.

Reply:
> Do you know why?

I don't. But it's been that way for a decade, I've not heard anyone complain ever, these members are legacy and discouraged from use, and this isn't a place to innovate (especially since there's no built-in set for which IsSynchronized is true). If you believe it's buggy and worth fixing, we can handle that for all of them as part of a separate issue you open, but this type shouldn't differ from those existing ones, and I'm not going to change them as part of this PR.

Reply:
Fair enough :)